### DISCUSSION TITLE
sparse-color Voronoi based on image (not a list of points)

### DISCUSSION THREAD

**Discussion by ranknfyle** [ImageMagick/ImageMagick#7861] [Help] [answered]
### ImageMagick version
7.1.1-43-Q8-x64

### Operating system, version and so on
Win 10

### Description
In the [Voronoi example](https://usage.imagemagick.org/canvas/#voronoi), the isolated colored points are supplied directly, on the command line. Is there a way to use an image as the source for the Voronoi pattern? The image is mostly black, with sparse. colored points (non-antialiased)

**Comment by fmw42** [comment]
Use txt: output and filter out all the black pixels. Then extract the non-black pixel colors and their coordinates from the txt: output.

**Reply by fmw42** [reply]
Or make your image with transparent black rather than opaque black. Then use that as in https://usage.imagemagick.org/canvas/#sparse_fill, but using Voronoi rather than Shepards

**Reply by ranknfyle** [reply]
Thanks for the suggestions. I think my image is too big for ImageMagick txt: approach, as I get a ~160 MB text file with pixel values. It will probably be wiser if I do this in code (Python etc).

**Comment by snibgo** [comment, marked as answer]
The piped text can contain just the non-transparent pixels. For example, assuming infile is 500x500 pixels, and black pixels are to be ignored for Voronoi:
```
magick infile.png -transparent Black sparse-color: | magick -size 500x500 xc:None -sparse-color Voronoi @- outfile.png
```

**Reply by ranknfyle** [reply]
> The piped text can contain just the non-transparent pixels.

Thanks, it works reasonably well, ~50 sec processing time, no text files. For the "-size" argument, can I somehow plug in the dimensions of infile.png?

**Comment by snibgo** [comment]
The size can come from a previous execution of magick.

Windows BAT:
```
for /F "usebackq" %%L in (`magick -ping infile.png -format "MYSIZE=%%wx%%h\n" info:`) do set %%L

magick infile.png -transparent Black sparse-color: | magick -size %MYSIZE% xc:None -sparse-color Voronoi @- outfile.png
```

Bash:
```
declare `magick -ping infile.png -format "MYSIZE=%wx%h\n" info:`

magick infile.png -transparent Black sparse-color: | magick -size ${MYSIZE} xc:None -sparse-color Voronoi @- outfile.png
```

I would like an enhancement to IM, so "-sparse-color" could use an image instead of text. The sparse list will often be generated by image processing, and having to convert this to text so IM can use it is awkward.

**Comment by fmw42** [comment]
This works for me.

Point Color Image (transparent background):

![sparse_source](https://github.com/user-attachments/assets/8f9d2183-e884-4b32-9109-1b819032a98c)


```
magick sparse_source.gif sparse-color:- |\
magick sparse_source.gif -alpha off \
-sparse-color voronoi '@-' sparse_voronoi.png
```

![sparse_voronoi](https://github.com/user-attachments/assets/5ea360b4-2bab-4733-a54c-3f4cb6075ecb)

**Comment by diamond-info** [comment]
I think the original message was about using this method in scripts or command-line mode with intermediate output to a temporary file. This example works for me:
 ```
magick sparse_source.gif sparse-color:- > "C:\temp\temp.txt"
magick sparse_source.gif -alpha off -colorspace RGB -sparse-color shepards "@C:\temp\temp.txt" sparse_voronoi.png
```
It's strange that this isn't given in the examples, but perhaps this method will be useful to someone..
